### PR TITLE
add capability to filter matching links in browser history

### PR DESCRIPTION
**Pull request by zhuopeng**
I received a freelancer task a few days ago that when extracting links using this plugin, also filter what's already in browser history. That's already been done, so thinking whether could this be a possible pull request to the main repository. 

Here's the brief changes I've done:
* add "history" to manifest to allow reading browser history
* add a new option "filter visited links", default to unchecked. 
* When the option is checked, read ALL chrome browser histories to filter against
* All test passed

The code changing is fairly straight forward, if you need further information, or further updating, please give feedback :)

Thanks!

### CODE REVIEW

File: src/background.js [hunk 1]
```diff
@@ -74,98 +74,121 @@ function timeConverter(a){
...
+	switch(request.message) {
+	case 'activate':
+		if(request.setting.options.block) {
+			request.urls = request.urls.unique();
+		}
+
+		if(request.setting.options.history){
+
+			// setting startTime to 0 and maxResults to Number.MAX_SAFE_INTEGER so to get every links in chrome history
+			chrome.history.search({text:"", startTime:0, maxResults:2^32-1}, function(items){
```
Comment (benblack86) [inline src/background.js]: Getting all history could be both very CPU and memory intensive. Could you instead use the search method to search for the particular URL?

**Comment by zhuopeng** [comment]
Thanks for review!

I'm not able to find a proper way to search for specific sets of URLs by using the search API

My initial idea is to use ```search``` with ```text``` as some sort of pattern or regular expression, after getting the search result, do a set minus from extracted URLs(by plugin) against matched URLs in chrome history, the result would be what we need. 

The obstacle now is how to perform the ```search``` which exactly matching all the provided links. I found no way to do that... After digging deep into the chrome history API, it use sql MATCH to query the history database
```c++
std::string sql = "SELECT url, title, time, offsets(pages), body FROM pages "
                    " LEFT OUTER JOIN info ON pages.rowid = info.rowid WHERE ";
  sql += options.body_only ? "body " : "pages ";
  sql += "MATCH ? AND time >= ? AND time < ? ORDER BY time DESC LIMIT ?";
  sql::Statement statement(db_.GetCachedStatement(SQL_FROM_HERE, sql.c_str()));

  // When their values indicate "unspecified", saturate the numbers to the max
  // or min to get the correct result.
  int64 effective_begin_time = options.begin_time.is_null() ?
      0 : options.begin_time.ToInternalValue();
  int64 effective_end_time = options.end_time.is_null() ?
      std::numeric_limits<int64>::max() : options.end_time.ToInternalValue();
  int effective_max_count = options.max_count ?
      options.max_count : std::numeric_limits<int>::max();

  statement.BindString(0, query);
  statement.BindInt64(1, effective_begin_time);
  statement.BindInt64(2, effective_end_time);
  statement.BindInt(3, effective_max_count);
```
I'm not sure how the ```MATCH``` works here, can't find any proper documentation around, but it looks like not a proper way of matching multiple value.

So, do you have any idea about this?

By the way, I tested in my browser, getting ALL history of size about 7k cost 150ms on a Macbook pro 2016. Haven't tested with larger history size though

if this is not possible, please just decline this pull request, should not be a general feature for performance risk

**Comment by benblack86** [comment]
I was thinking it might be better to iterate over each url and call `chrome.history.search` on each, but maybe that will take longer depending on the latency of `chrome.history.search`.

**Comment by zhuopeng** [comment]
Tried to check how chrome handle ```chrome.history.search``` API call. Here are 
* use sqlite to store history
* when performing API call, two paths, if the ```text``` parameter is empty, perform a get all, else if ```text``` have value, perform full text searching by using ```MATCH```
* sqlite ```MATCH``` support an ```OR``` operator which works as the name suggest, https://sqlite.org/fts3.html#set_operations_using_the_standard_query_syntax. Which seems to be a work around to query multiple URLs at a time. However, Chrome inserted a query parser in the middle which tokenize the ```text``` value and do its magic. For example query of ```com OR org``` then transformed to ```com* OR* org*```, which failed my purpose... 

That conclude several things:
1. search history with set of URLs is not possible
1. on each URL to perform a ```chrome.history.search``` will be much slower
2. query All URLs(with ```text``` set to empty) is slightly faster than any valued search. I tested the truthiness on my laptop on this, because full text searching need to scan all rows as well.  

So, I suppose the best solution right now would be the initial submit. Only for the RAM usage if history is huge. But this option is disabled by default, and in the latest commit, I added a warning

Again, please decide whether this feature is needed :)

**Comment by benblack86** [comment]
Due to performance concerns, I will close the PR. It sounds like a useful features, but I would prefer if the built in search functionality performed better.

**Pull request closed without merging**